massa-async-pool/src/pool.rs

```diff
@@ -0,0 +1,201 @@
+    /// and get the list of pruned messages for `coins` reimbursement.
+    ///
+    /// # arguments
+    /// * slot: used to filter out expired messages, not stored
+    /// * new_messages: list of AsyncMessage to add to the pool
+    ///
+    /// # returns
+    /// The list of (message_id, message) that were eliminated from the pool after the changes were applied, sorted in the following order:
+    /// * expired messages from the pool, in priority order (from lowest to highest priority)
+    /// * expired messages from new_messages (in the order they appear in new_messages)
+    /// * excess messages after inserting all remaining new_messages, in priority order (from lowest to highest priority)
+    pub fn settle_slot(
+        &mut self,
+        slot: Slot,
+        new_messages: Vec<AsyncMessage>,
+    ) -> Vec<(AsyncMessageId, AsyncMessage)> {
+        // Compute IDs
+        let mut new_messages: Vec<_> = new_messages
+            .into_iter()
+            .map(|v| (v.compute_id(), v))
+            .collect();
+
+        // Filter out all messages for which the validity end is expired.
+        // Note that the validity_end bound is NOT included in the validity interval of the message.
+        let mut eliminated: Vec<_> = self
+            .messages
+            .drain_filter(|_k, v| slot >= v.validity_end)
+            .chain(new_messages.drain_filter(|(_k, v)| slot >= v.validity_end))
+            .collect();
+
+        // Insert new messages into the pool
+        self.messages.extend(new_messages);
+
+        // Truncate message pool to its max size, removing non-prioritary items
+        let excess_count = self
+            .messages
+            .len()
+            .saturating_sub(self.config.max_length as usize);
+        eliminated.reserve_exact(excess_count);
+        for _ in 0..excess_count {
+            eliminated.push(self.messages.pop_first().unwrap()); // will not panic (checked at excess_count computation)
+        }
+
+        eliminated
+    }
+
+    /// Takes the best possible batch of messages to execute, with gas limits and slot validity filtering.
+    /// The returned messages are removed from the pool.
+    /// This method is used at the beginning of a slot execution to list async messages to execute.
+    ///
+    /// # arguments
+    /// * slot: select only messages that are valid within this slot
+    /// * available_gas: maximum amount of available gas
+    ///
+    /// # returns
+    /// A vector of messages, sorted from the most prioritary to the least prioritary
+    pub fn take_batch_to_execute(
+        &mut self,
+        slot: Slot,
+        mut available_gas: u64,
+    ) -> Vec<AsyncMessage> {
+        // gather all selected items and remove them from self.messages
+        // iterate in decreasing priority order
+        let (removed, kept) = std::mem::take(&mut self.messages)
+            .into_iter()
+            .rev()
+            .partition(|(_, msg)| {
```

Comment:
I have one concern with partition() => will the "kept" part be completely re-sorted ?

Reply:
Why not use drain_filter ?

Reply:
Because we have to `rev` before which need an iterator and then drain filter that don't work on iterator and we can't collect object reversed in a btreemap because they are sorted directly.

Reply:
Partition will keep the order so it will not need a re-sorting. I don't know if it's more costly or not

Reply:
in the end, we chose to keep objects in reverse order (using https://doc.rust-lang.org/std/cmp/struct.Reverse.html ) and do drain_filter i